test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala

```diff
@@ -0,0 +1,236 @@
+package scala.scalanative
+package testinterface
+
+import java.io.{
+  ByteArrayInputStream,
+  ByteArrayOutputStream,
+  DataInputStream,
+  DataOutputStream
+}
+
+import scala.compat.Platform.EOL
+import scala.scalanative.posix._
+import scala.scalanative.native._
+import scala.scalanative.runtime.ByteArray
+import arpa.inet._
+import sys.socket
+import netinet.in._
+import sbt.testing.{Event => SbtEvent, _}
+
+import scala.scalanative.testinterface.serialization._
+import Serializer._
+import scala.annotation.tailrec
+import scala.scalanative.posix.inttypes.uint16_t
+import scala.scalanative.posix.sys.socket.{accept, sockaddr, socklen_t}
+
+abstract class TestMainBase {
+
+  /** All the frameworks reported in `loadedTestFrameworks` in sbt. */
+  def frameworks: Seq[Framework]
+
+  /** A mapping from class name to instantiated test object. */
+  def tests: Map[String, AnyRef]
+
+  /** Actual main method of the test runner. */
+  def testMain(args: Array[String]): Unit = Zone { implicit z =>
+    val server_port = args.headOption.map(_.toInt).getOrElse(9000).toUShort
+    val listen_sock = setupServer(server_port)
+
+    onClient(listen_sock) { client_socket =>
+      // We don't need to listen for incoming connections anymore
+      unistd.close(listen_sock)
+      testRunner(Array.empty, null, client_socket)
+    }
+  }
+
+  /** Executes body, calls `perror` with `str` if the result is non-zero. */
+  private def exitOnFailure(str: String)(body: => CInt)(
+      implicit zone: Zone): Unit = {
+    val err = body
+    if (err != 0) {
+      stdio.perror(toCString(str))
+      scala.sys.exit(err)
+    }
+  }
+
+  /** Sets up the server socket, returns the file descriptor. */
+  private def setupServer(port: uint16_t)(implicit zone: Zone): CInt = {
+    val server_address = native.alloc[sockaddr_in]
+    !server_address._1 = socket.AF_INET.toUShort
+    !server_address._2 = htons(port)
+    !server_address._3._1 = htonl(INADDR_ANY)
```

Comment:
sockaddr and friends badly need to have implicit classes that alias their field names to their human-readable counter-parts.